lkl: add new host_ops, seterrno() to notify errno to applications

thehajime opened the pull request:
This will change the return value of LKL system call (i.e., lkl_sys_xxx)
from returning error number of sys_xxx() call directly to -1 with
setting errno with the error number in thread-local-storage (i.e., errno).

Without that, both LKL system calls (lkl_sys_recvmsg) and standard
system calls (recvmsg(2)) have different specification of return value
although other parts are almost the same between them. We may probably
need duplicate documentations between them as well.

This also needs to update the test programs (boot.c) so that it involves
the different API.

Signed-off-by: Hajime Tazaki thehajime@gmail.com

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/lkl/linux/5)

<!-- Reviewable:end -->

cemeyer commented:
I think this is maybe the wrong layer for this.  The kernel code shouldn't change to set errno; instead the syscall wrappers should convert the kernel error statuses into userspace -1/errno conventions, like libc does today.

thehajime commented:
@cemeyer thanks for the comment. it makes sense.

then I have options to address this: comments are again so much welcome.

1) move this err code translation lkl_sys_xxx() (treat lkl_sys_xxx a libc call)
we will provide a single API to users, but change the current LKL system call API.

2) introduce lkl_sys_xxx_wrapper() (introduce libc-alike layer)
I will introduce this by an external script to automatically generate this.

we will have two API to (LKL) users
- LKL system call (lkl_sys_xxx)
- (libc-alike) system call (lkl_sys_xxx_wrapper, name suggestion welcome)

any thought ?

cemeyer commented:
Do we really need or want it?  lkl_sys routines return *LKL* errors, not host kernel error values.  IMO it doesn't make sense to assign `LKL_EFOO` to the host libc's `errno` variable.  On a Linux host system, the LKL errnos will *probably* match up with the host errnos (purely by accident), but on NT and FreeBSD hosts they will not match.

thehajime commented:
@cemeyer I was initially thinking about existing Linux userspace app + LKL, where this errno translation is useful. For only this case, `LKL_EFOO` assign directly to host `errno` make sense and will ease the porting process to LKL-aware apps.

If you want map error values in foreign host OSs, you can do in `host_ops->seterrno()`: that's why I outsourced this part.

thehajime commented:
I reworked the errno handling bits.

This version went to avoid changing error value inside lkl_syscall(). instead changing the value at another functions.

I was thinking to add the errno handling in external files but having similar definitions here and there may puzzled (at least for me) so, I put additional lines into LKL_SYSCALL macro.

I tentatively named the functions by lkl_sys_impl_xxx(), but any rename suggestions are more than welcome (e.g., lkl_sys_wrapper_xxx(), etc).

thehajime commented:
I fixed typos and rebased the branch. thanks @cemeyer.

thehajime commented:
rebased with the latest master.

> I tentatively named the functions by lkl_sys_impl_xxx(), but any rename suggestions are more than welcome (e.g., lkl_sys_wrapper_xxx(), etc).

I changed the name this API to lkl_sys_wrapper_xxx().

tavip commented:
I will not merge this. This is only useful for system call hijacking and it can be implemented there instead of having it done at LKL level.

thehajime commented:
acked.
